### DISCUSSION TITLE
Web browsing using # isn't scraping content

### DISCUSSION THREAD

**Discussion by rfernandez760** [open-webui/open-webui#6189] [General]
# Bug Report

**Important: Before submitting a bug report, please check whether a similar issue or feature request has already been posted in the Issues or Discussions section. It's likely we're already tracking it. In case of uncertainty, initiate a discussion post first. This helps us all to efficiently focus on improving the project.**

**Let's collaborate respectfully. If you bring negativity, please understand our capacity to engage may be limited. If you're open to learning and communicating constructively, we're more than happy to assist you. Remember, Open WebUI is a volunteer-driven project maintained by a single maintainer, supported by our amazing contributors who also manage full-time jobs. We respect your time; please respect ours. If you have an issue, We highly encourage you to submit a pull request or to fork the project. We actively work to prevent contributor burnout to preserve the quality and continuity of Open WebUI.**

## Installation Method

OWUI Installed via Docker

## Environment

- **Open WebUI Version:** v0.3.32
- **Ollama (if applicable):** 0.3.13

- **Operating System:** Win11, Running WSL2 with Ubuntu
- **Browser (if applicable):** Chrome and Edge

**Confirmation:**

- [x] I have read and followed all the instructions provided in the README.md.
- [x] I am on the latest version of both Open WebUI and Ollama.
- [ ] I have included the browser console logs.
- [x] I have included the Docker container logs.
- [x] I have provided the exact steps to reproduce the bug in the "Steps to Reproduce" section below.

## Expected Behavior:

Scenario 1. When adding in #<URL> at the beginning or in the message, OWUI will generate a file for the LLM review much like attaching a file that has been uploaded for RAG.
Scenario 2. When adding #<URL> in the message, OWUI will scrape the content of the web page for RAG. Regardless of the type of LLM model selected.

## Actual Behavior:

Scenario 1. The model is not able to retrieve any data. The citation just shows the URL link that is provided. 
Scenario 2. Closed source models like Claude or GPT4.0 can't read the web page, but open source models can. Please note the open source models tested were llama3.1 that is installed locally as well as open source models connected to Groq via API.

## Description

**Bug Summary:**
Web browsing (ability to read a web page) is unavailable, and inconsistent behavior based on open source vs closed source models. This I don't believe is the expected functionality.

## Reproduction Details

**Steps to Reproduce:**
Scenario 1: Add a #<URL> first in the chat so OWUI can generate the document (content) for review. Then ask any questions like summarize. Use any model open source or closed source.

Scenario 2: Type in the message field "Please summarize the following #<URL>". Ask using an open source, and then create a new message following the same steps using a closed source model. The open is able to retrieve and read, while the closed source doesn't have any context.

Example URL: #[https://www.msn.com/en-us/news/politics/cruz-calls-out-mcconnell-aligned-group-for-not-spending-a-penny-in-texas/ar-AA1sgaor?ocid=msedgntp&amp;pc=U531&amp;cvid=8e102387764f4ce3925621b1d3a529c0&amp;ei=10](https://www.msn.com/en-us/news/politics/cruz-calls-out-mcconnell-aligned-group-for-not-spending-a-penny-in-texas/ar-AA1sgaor?ocid=msedgntp&pc=U531&cvid=8e102387764f4ce3925621b1d3a529c0&ei=10)

## Logs and Screenshots

**Browser Console Logs:**
[Include relevant browser console logs, if applicable]

**Docker Container Logs:**
[Include relevant Docker container logs, if applicable]
[Docker Logs.txt](https://github.com/user-attachments/files/17369905/Docker.Logs.txt)

**Screenshots/Screen Recordings (if applicable):**
![image](https://github.com/user-attachments/assets/9a78abd0-932e-4b94-a779-04d8a639109f)

![image](https://github.com/user-attachments/assets/797d6f68-bd12-4716-96de-ff5d57d16b7e)

## Additional Information

[Include any additional details that may help in understanding and reproducing the issue. This could include specific configurations, error messages, or anything else relevant to the bug.]

## Note

If the bug report is incomplete or does not follow the provided instructions, it may not be addressed. Please ensure that you have followed the steps outlined in the README.md and troubleshooting.md documents, and provide all necessary information for us to reproduce and address the issue. Thank you!

**Comment by tjbck** [comment]
This occurs due to this specific website blocking the web scraping. You can test this by using the web scraping feature using our docs website: https://docs.openwebui.com/

**Reply by rfernandez760** [reply]
oh, gotcha. I did just verify it working for a page that G:30b just used in a reply. Any ideas as to why Claude and other closed source LLMs can't see it?

**Reply by bsima** [reply]
If a website is blocking the web scraping, and error should be thrown and presented in the UI somehow. Otherwise the llm just hallucinates a web page.

**Reply by biebiep** [reply]
> This occurs due to this specific website blocking the web scraping. You can test this by using the web scraping feature using our docs website: https://docs.openwebui.com/

@tjbck, I don't think this is correct.
I didn't want to open a new bug as this one exists. This issue still exists in v0.6.22 today.

Page content is being scraped correctly, as I can open the doc and see the page content.

Error is as follows, just as described in above bug report.
a) When using Ollama, the model is able to see the page content as well and you are able to interact with it.
b) But if I use external Azure OpenAI API, the context doesn't get sent with the API call somehow and the model states it's unaware.

Situation b) is definitely broken.

**Reply by biebiep** [reply]
@tjbck the converter from OpenAI to Ollama seems to fix a bug in the retrieval flow as a "happy side effect", the hashtag based scraping flow is, i think, missing a key facet.

The filetype of a "scraped object" is "doc". The retrieval/utils.py handler doesn't handle files of type "doc".  (see here) https://github.com/open-webui/open-webui/blob/438e5d966f0f64f9ea3feab22724a5bd96a4127b/backend/open_webui/retrieval/utils.py#L484
It does handle an array of "docs" for web search, but not the singular case.

So when you're using OpenAI API's, the docs seemingly get ignored because their filetype isn't handled.

Either I change the filetype metadata from the scraper from doc to text (haven't identified where yet in the code), or I change the  retrieval handler to handle "doc" filetypes the same as like the "text" workflow. (this seemed to work in testing and didn't break the Ollama compatibility)

**Reply by rgaricano** [reply]
I think that this objects (doc) should be managed as files, in https://github.com/open-webui/open-webui/blob/438e5d966f0f64f9ea3feab22724a5bd96a4127b/backend/open_webui/retrieval/utils.py#L522

**Reply by biebiep** [reply]
https://github.com/open-webui/open-webui/pull/16736

I treated them as flat text, as that is what the scraper returns. Tested and works against Ollama and OpenAI endpoints.


I agree it's ambiguous with "file", but I don't think there is a physical download of the page content to the /uploads/data folder? When there is no file on disk, I think the "file" moniker is misleading. I chose text because the documentation of that "text" component said it was for local chat uploads, which IMO is what the # functionality does.

**Comment by rfernandez760** [comment]
@tjbck thanks for all the great work. Just wanted to let you know that whatever update you and the devs recently rolled out has fixed / enabled the capability again with third party LLMs.

**Comment by rgaricano** [comment]
working fine here:

![openwebui_test_weburl](https://github.com/user-attachments/assets/187b866c-2e2c-4f90-9c3c-656b735ba799)

![openwebui_test_weburl1](https://github.com/user-attachments/assets/ff108992-1cf7-499d-9321-46fff28825a5)

**Reply by biebiep** [reply]
As stated in the bug report  "Scenario 2" clearly: 
**local models based on ollama work**. As Cogito is on HF.co, I'm going to assume you're hosting it yourself.
OpenAI based models (for instance Azure OpenAI) do not receive the scraped data.

So again;
The web scraping itself _works_ as intended on both situations; verified because you can read the content of the site in the document popup. The issue is that somehow when running through ollama the text is passed to the LLM, but passing it to Azure GPT4o, for instance, doesn't work. 

My "quick fix" that fixes it on our platform above would suggest it's because the scraped url is passed as a item.type==doc and there is no handler for that.  Web search is passed as list in item.docs so that does have a handler/ does work as you point out in my pull request.

My best guess for the reason for this discrepancy is likely due to how the chat.py converts the openai payload to an ollama payload.


I don't know how to spell it out more clearly than has already been done in the OP of this thead. You clearly show a local model working, which we all claim is working. Try it with Azure OpenAI endpoints. It fails to pass the context.

**Reply by rgaricano** [reply]
sorry about ny misconfussion,
the conversion is ollama to openai not openai to ollama.

**Reply by biebiep** [reply]
I am actually unsure how your example works. For me when I add # url, it actually immediately presents itself as a document attachment, not as the url like in your interface.

**Reply by rgaricano** [reply]
~~Ok, that it's because I enter it at begin of the message,
When it's entered not at the begin it is detected as file/document and, in my case, arise an error toast of empty content...
for some reason don't do scrap&embedd in this case.~~

EDIT:
I test again and now it work fine, adding the `#https://www.msn.com/en-us/news/politics/cruz-calls-out-mcconnell-aligned-group-for-not-spending-a-penny-in-texas/ar-AA1sgaor?ocid=msedgntp&pc=U531&cvid=8e102387764f4ce3925621b1d3a529c0&ei=10`  command in any place,
also I recheck the logic and it's the same, it is detected as web in [src/lib/components/chat/MessageInput/Commands.svelte#L87](https://github.com/open-webui/open-webui/blob/438e5d966f0f64f9ea3feab22724a5bd96a4127b/src/lib/components/chat/MessageInput/Commands.svelte#L87) and proccessed in [src/lib/components/chat/MessageInput/Commands/Knowledge.svelte#L324](https://github.com/open-webui/open-webui/blob/438e5d966f0f64f9ea3feab22724a5bd96a4127b/src/lib/components/chat/MessageInput/Commands/Knowledge.svelte#L324), the validation of the url is done in [src/lib/utils#L786](https://github.com/open-webui/open-webui/blob/438e5d966f0f64f9ea3feab22724a5bd96a4127b/src/lib/utils/index.ts#L786)

**Reply by biebiep** [reply]
You're correct of course, but I don't see how it's related to the issue where the uploaded file is only passed to the model when it's local/ollama based and not when its OpenAI based.

**Reply by rgaricano** [reply]
It's fine for you it I use a openAI API compatible connection?
![openwebui_oll_oai_comp](https://github.com/user-attachments/assets/219536fc-314f-4f56-ae07-61422980a9a3)
![openwebui_oll_oai_comp1](https://github.com/user-attachments/assets/a515cf97-6544-4857-aa8a-127728e2600a)
![openwebui_oll_oai_comp2](https://github.com/user-attachments/assets/7f384ace-8112-4be3-ac87-18ead5a2ce4b)


because when I try to run openAI it just wait waiting...until I refresh.
![openwebui_oll_oai_comp3](https://github.com/user-attachments/assets/c3f5af1e-804f-4d61-b02b-8bd76fddd8ad)
![openwebui_oll_oai_comp4](https://github.com/user-attachments/assets/8465a5db-c24f-4d28-931b-ee6ffd20a0c7)
![openwebui_oll_oai_comp5](https://github.com/user-attachments/assets/a318f46c-4d77-484c-aa06-71f8d975be0b)

In my case, I think that this error is not well catched and break the renderer: _429 - You exceeded your current quota, please check your plan and billing details_

**Reply by biebiep** [reply]
I still don't understand why your hashtag behaviour is so fundamentally different from mine.  Do you just type it out and not click on the popup or not press Tab to complete the action after pasting the url?

Where is your attachment icon? OWUI automatically removes the url from the message for me and places the website as an attachment in the conversation. Sorry for the lack of screenshots, I'll see if I can add some tomorrow.

**Reply by rgaricano** [reply]
nothing, I just paste it and send,
my configs are the ones I posted, & in user interface _Web Search in Chat_ is set as _Default_
